src/go/k8s/controllers/redpanda/cluster_controller_configuration.go

```diff
@@ -0,0 +1,275 @@
+		_, err := adminAPI.PatchClusterConfig(patch.Upsert, patch.Remove)
+		if err != nil {
+			return fmt.Errorf("could not patch centralized configuration on cluster %s/%s: %w", redpandaCluster.Namespace, redpandaCluster.Name, err)
+		}
+	}
+
+	// TODO a failure and restart here may lead to inconsistency if the user changes the CR in the meantime (e.g. removing a field),
+	// since we applied a config to the cluster but did not store the information anywhere else.
+	// A possible fix is doing a two-phase commit (first stage commit on configmap, then apply it to the cluster, with possibility to recover on failure),
+	// but it seems overkill given that the case is rare and requires cooperation from the user.
+
+	hash, err := config.GetCentralizedConfigurationHash(schema)
+	if err != nil {
+		return fmt.Errorf("could not compute hash of the new configuration for cluster %s/%s: %w", redpandaCluster.Namespace, redpandaCluster.Name, err)
+	}
+
+	oldHash, err := statefulSetResource.GetCentralizedConfigurationHashFromCluster(ctx)
+	if err != nil {
+		return err
+	}
+
+	if oldHash == "" {
+		// Annotation not yet set on the statefulset (e.g. first time we change config).
+		// We check a diff against last applied configuration to avoid triggering a restart when not needed.
+		prevConfig := *config
+		prevConfig.ClusterConfiguration = lastAppliedConfiguration
+		oldHash, err = prevConfig.GetCentralizedConfigurationHash(schema)
+		if err != nil {
+			return err
+		}
+	}
+
+	if oldHash != hash {
+		// Definitely needs restart
+		if err := statefulSetResource.SetCentralizedConfigurationHashInCluster(ctx, hash); err != nil {
+			return fmt.Errorf("could not update config hash on statefulset %s/%s: %w", redpandaCluster.Namespace, redpandaCluster.Name, err)
+		}
+	}
+
+	// Mark the new lastAppliedConfiguration for next update
+	if err := configMapResource.SetLastAppliedConfigurationInCluster(ctx, config.ClusterConfiguration); err != nil {
+		return fmt.Errorf("could not store last applied configuration in the cluster: %w", err)
+	}
+
+	// Check status again using admin API
+	status, err = adminAPI.ClusterConfigStatus()
+	if err != nil {
+		return fmt.Errorf("could not get config status from admin API")
+	}
+	originalClusterStatus := redpandaCluster.Status.DeepCopy()
+	originalCondition := originalClusterStatus.GetCondition(redpandav1alpha1.ClusterConfiguredConditionType)
+	conditionData := mapToCondition(status)
+	redpandaCluster.Status.SetCondition(conditionData.Type, conditionData.Status, conditionData.Reason, conditionData.Message)
+	currentCondition := redpandaCluster.Status.GetCondition(redpandav1alpha1.ClusterConfiguredConditionType) // contains time wrt conditionData
+	stsNeedsRestart := needsRestart(status)
+	if !reflect.DeepEqual(originalCondition, currentCondition) || (stsNeedsRestart && !redpandaCluster.Status.Upgrading) {
+		// Trigger restart here if needed
+		if stsNeedsRestart {
+			redpandaCluster.Status.Upgrading = true
```

Comment:
Is the assumption here that rolling restarts are going to be totally safe and transparent by the time this is released?  Otherwise it's a bit risky to auto-restart on config edits.

Also, it seems like the "Upgrading" field should be renamed if it is also used to trigger restarts generally

Reply:
Afaik auto-restart is the default behavior even without these changes. Is this correct @RafalKorepta ?

The upgrading field is misleading, right. Something like `restarting` (if a field is needed) would be better. I started to work on rolling restarts so I may fix the field in that PR.